### PR TITLE
Go: Improve QHelp for `go/unsafe-quoting`.

### PR DESCRIPTION
**Pull request by max-schaefer**

### CODE REVIEW

**Review by owen-mc — APPROVED** [review]

File: go/ql/src/Security/CWE-089/StringBreak.qhelp [hunk 1]
```diff
@@ -5,19 +5,21 @@
...
-data embedded in between the quotes does not itself contain a quote. Otherwise the embedded data
-could (accidentally or intentionally) change the structure of the overall string by terminating
-the quoted substring early, with potentially severe consequences. If, for example, the string is
-later interpreted as an operating-system command or database query, a malicious attacker may be
-able to craft input data that enables a command injection or SQL injection attack.
+Code that constructs a quoted string literal containing user-provided data needs to ensure that
+this data does not itself contain a quote. Otherwise the embedded data could (accidentally or
+intentionally) terminate the string literal early and thereby change the structure of the overall
+string, with potentially severe consequences. If, for example, the string is later used as
+part an operating-system command or database query, an attacker may be able to craft input data
```
Comment (felicitymay) [inline go/ql/src/Security/CWE-089/StringBreak.qhelp:12]: Missing word?

```suggestion
part of an operating-system command or database query, an attacker may be able to craft input data
```

File: go/ql/src/Security/CWE-089/StringBreak.qhelp [hunk 2]
```diff
@@ -5,19 +5,21 @@
...
+that injects a malicious command.
 </p>
 </overview>
 
 <recommendation>
 <p>
 Sanitize the embedded data appropriately to ensure quotes are escaped, or use an API that does
-not rely on manually constructing quoted substrings.
+not rely on manually constructing quoted substrings. Make sure to use the appropriate escaping
+mechanism, for example double quoting for SQL strings or backslash escaping for shell commands.
```
Comment (felicitymay) [inline go/ql/src/Security/CWE-089/StringBreak.qhelp:21]:
```suggestion
mechanism, for example, double quoting for SQL strings or backslash escaping for shell commands.
```

File: go/ql/src/Security/CWE-089/StringBreak.qhelp [hunk 3]
```diff
@@ -29,17 +31,38 @@ then embeds it into a SQL query built using the Squirrel library.
...
 </p>
 <p>
-To fix this vulnerability, use Squirrel's placeholder syntax, which avoids the need to explicitly
-construct a quoted string.
+To fix this vulnerability, use the placeholder syntax from Squirrel's structured API for building
+queries, which avoids the need to explicitly construct a quoted string.
 </p>
 <sample src="StringBreakGood.go"/>
+<p>
+In situations where a structured API is not available, make sure to escape quotes before embedding
```
Comment (felicitymay) [inline go/ql/src/Security/CWE-089/StringBreak.qhelp:45]:
```suggestion
In situations where a structured API is not available, make sure that you escape quotes before embedding
```

File: go/ql/src/Security/CWE-089/StringBreak.qhelp [hunk 4]
```diff
@@ -29,17 +31,38 @@ then embeds it into a SQL query built using the Squirrel library.
...
+user-provided data into a quoted string. For example, this is how you can backslash-escape single
+quotes using <code>strings.ReplaceAll</code>:
+</p>
+<sample language="go">
+  quoted := strings.ReplaceAll(raw, `\`, `\\`)
+  quoted = strings.ReplaceAll(quoted, "'", "\\'")
+</sample>
+<p>
+Note that any existing backslash characters in the string must be escaped first, so that they do
```
Comment (felicitymay) [inline go/ql/src/Security/CWE-089/StringBreak.qhelp:54]: Would it make sense to update the example above to include both types of escaping, or is this totally obvious to the intended audience?

**Review by felicitymay — DISMISSED** [review]
A few small comments, otherwise looks good to me ✨

File: go/ql/src/Security/CWE-089/StringBreak.qhelp (hunk 4 above)
Comment (max-schaefer) [reply go/ql/src/Security/CWE-089/StringBreak.qhelp:54]: When you say both types of escaping, are you referring to backslash escaping and doubling quotes? I think that's fairly obvious to our audience.

**Review by felicitymay — APPROVED** [review]

**Pull request merged**